### PR TITLE
Enable shellcheck for external test scripts

### PR DESCRIPTION
**Pull request by cameel**
**This PR depends on #10468 and should be merged after it.**

I noticed that we have a big shellcheck ignore list which includes these scripts among other things. Here's a small step towards removing it.

### CODE REVIEW

**Comment by hrkrshnn** [comment]
Is this change needed for the tests or is this just a refactor?

File: scripts/chk_shellscripts/ignore.txt [hunk 1]
```diff
@@ -1,12 +1,5 @@
 ./test/docsCodeStyle.sh
 ./test/cmdlineTests.sh
-./test/externalTests.sh
-./test/externalTests/common.sh
-./test/externalTests/gnosis.sh
-./test/externalTests/zeppelin.sh
-./test/externalTests/colony.sh
-./test/externalTests/solc-js/solc-js.sh
-./scripts/common.sh
```
Comment (bshastry) [inline scripts/chk_shellscripts/ignore.txt:9]: Ah, I wondered why shellcheck ran only for ens.sh and not others. This explains it!

File: test/externalTests/colony.sh [hunk 2]
```diff
@@ -31,18 +31,17 @@ function test_fn { yarn run test:contracts; }
 function colony_test
 {
     OPTIMIZER_LEVEL=3
-    FORCE_ABIv2=false
     CONFIG="truffle.js"
 
-    truffle_setup https://github.com/solidity-external-tests/colonyNetwork.git develop_070
-    run_install install_fn
+    truffle_setup "$SOLJSON" https://github.com/solidity-external-tests/colonyNetwork.git develop_070
```
Comment (bshastry) [inline test/externalTests/colony.sh:36]: This looks like a non shellcheck related change, does it belong in this PR?

File: test/externalTests/common.sh [hunk 3]
```diff
@@ -232,11 +235,15 @@ function run_test
...
 function truffle_run_test
 {
-    local compile_fn="$1"
-    local test_fn="$2"
+    local soljson="$1"
+    local compile_fn="$2"
+    local test_fn="$3"
+    local force_abi_v2_flag="$4"
+
+    test [ "$force_abi_v2_flag" = force_abi_v2 ] || [ "$force_abi_v2_flag" = no_force_abi_v2 ]
```
Comment (bshastry) [inline test/externalTests/common.sh]: More non shellcheck related changes here

File: test/externalTests/ens.sh [hunk 4]
```diff
@@ -18,9 +18,7 @@
 #
 # (c) 2019 solidity contributors.
 #------------------------------------------------------------------------------
-# shellcheck disable=SC1091
 source scripts/common.sh
-# shellcheck disable=SC1091
```
Comment (bshastry) [inline test/externalTests/ens.sh:23]: Nice!

**Review by bshastry — COMMENTED** [review]
Looks mostly good to me. Could the non shellcheck related changes go into a new PR?

File: test/externalTests/colony.sh (hunk 2 above)
Comment (cameel) [reply test/externalTests/colony.sh:36]: Shellcheck complains about `SOLJSON` being potentially uninitialized - because it's supposed to be set by the script and then used by `common.sh`.

So it's related, it's just not the most minimal fix I could do :) I could have just added some dummy initialization at the top of the script but these globals make the external test scripts hard to follow for me so I took the occasion to eliminate some of them.

File: test/externalTests/common.sh (hunk 3 above)
Comment (cameel) [reply test/externalTests/common.sh]: Similar to https://github.com/ethereum/solidity/pull/10464#discussion_r534117440. Shellcheck complained about `FORCE_ABIv2` being an uninitialized global.

**Comment by cameel** [comment]
@hrkrshnn 
> Is this change needed for the tests or is this just a refactor?

It's not needed for tests to work but I added a new script and the shellcheck CI job started failing. I had a choice of either adding the script to shellcheck ignore list or fixing the stuff shellcheck reports. I decided to do the latter so in a way this PR is a fix needed for #10429 to pass all tests - but again not the most minimal possible fix. If there's an actual need I can always rebase #10429 directly on `develop` and just add the ignore instead.

File: test/externalTests/colony.sh (hunk 2 above)
Comment (bshastry) [reply test/externalTests/colony.sh:36]: Does shellcheck complain about https://github.com/koalaman/shellcheck/wiki/SC2154?

Comment (cameel) [reply test/externalTests/colony.sh:36]: Yes. Exactly.

Comment (cameel) [reply test/externalTests/colony.sh:36]: Ah, wait. I think that `SC2154` was the error for `FORCE_ABIv2`. For this one it was that `soljson` command exists and `SOLJSON` looks like a wrongly capitalized version of it or something like that. That's why it only complained about these two and was fine with `DIR` and the other fully capitalized globals.

File: test/externalTests/common.sh [hunk 5]
```diff
@@ -92,10 +93,11 @@ function force_truffle_version
 
 function truffle_setup
 {
-    local repo="$1"
-    local branch="$2"
+    local soljson="$1"
```
Comment (bshastry) [inline test/externalTests/common.sh:96]: Note to self: truffle_setup now accepts three inputs instead of two

File: test/externalTests/common.sh [hunk 6]
```diff
@@ -207,11 +209,12 @@ function clean
 
 function run_install
 {
-    local init_fn="$1"
+    local soljson="$1"
```
Comment (bshastry) [inline test/externalTests/common.sh:212]: run_install accepts two inputs instead of one

File: test/externalTests/common.sh [hunk 7]
```diff
@@ -40,9 +40,10 @@ function verify_version_input
 
 function setup
 {
-    local branch="$1"
+    local soljson="$1"
+    local branch="$2"
```
Comment (bshastry) [inline test/externalTests/common.sh:44]: setup accepts two inputs instead of one

File: test/externalTests/common.sh [hunk 8]
```diff
@@ -232,11 +235,15 @@ function run_test
 
 function truffle_run_test
 {
-    local compile_fn="$1"
-    local test_fn="$2"
+    local soljson="$1"
+    local compile_fn="$2"
+    local test_fn="$3"
+    local force_abi_v2_flag="$4"
```
Comment (bshastry) [inline test/externalTests/common.sh:241]: truffle_run_test accepts four inputs instead of two

File: test/externalTests/common.sh (hunk 3 above)
Comment (bshastry) [reply test/externalTests/common.sh]: Since `force_abi_v2` is also the name of a function, could you please rename it for readability?

Comment (bshastry) [reply test/externalTests/common.sh]: Also, I'm not sure what the `test` achieves. Is it meant to silence a shellcheck warning by making use of something that was previously initialised?

File: test/externalTests/common.sh [hunk 9]
```diff
@@ -258,7 +265,7 @@ function truffle_run_test
         clean
         force_solc_settings "$CONFIG" "$optimize" "istanbul"
         # Force abi coder v2 in the last step. Has to be the last because code is modified.
-        if [ "$FORCE_ABIv2" = true ]; then
+        if [ "$force_abi_v2_flag" = force_abi_v2 ]; then
```
Comment (bshastry) [inline test/externalTests/common.sh]: Could this be retained as a boolean?

**Review by bshastry — CHANGES_REQUESTED** [review]
Mostly looks good. Only suggestion I have is to refactor the force abiv2 variable e.g., rename so it doesn't shadow existing function of the same name and make it a boolean instead of string

File: test/externalTests/common.sh (hunk 5 above)
Comment (cameel) [reply test/externalTests/common.sh:96]: I'm assuming this won't a big problem because the use of these functions is pretty localized. If I were attempting to use them I'd copy and existing call and modify it anyway.

File: test/externalTests/common.sh (hunk 9 above)
Comment (cameel) [reply test/externalTests/common.sh]: I specifically did not want this to be a boolean to make the call more self-explanatory. I prefer this:

```bash
truffle_run_test "$SOLJSON" compile_fn test_fn no_force_abi_v2
```

to this:
```bash
truffle_run_test "$SOLJSON" compile_fn test_fn false
```

I also try to avoid using `true` and `false` as booleans in Bash when I have a better alternative because it can easily bite you in some contexts. Especially the fact that `if [ $var = true ]` and `if [ $var = false ]` will just silently do the wrong thing when `var` has a value other than `true` or `false`.

Admittedly, `no_force_abi_v2` looks too much like a function name in this context. I'll rename it to something else.

By the way, no test is currently using this flag so I could also just remove it.

File: test/externalTests/common.sh (hunk 3 above)
Comment (cameel) [reply test/externalTests/common.sh]: Good point. I'll rename `force_abi_v2`. I was struggling to find something better but maybe just using something like `force-abi-v2` will make this more apparent.

The test is just an assertion/validation. It ensures that you'll get an error (due to `set -e` at the top of the file) if you try to pass anything other than `force_abi_v2` or `no_force_abi_v2` in the `force_abi_v2_flag` parameter. Since I'm changing the signature of the function and Bash is very lax about parameters, I wanted to have an additional sanity check in case I messed this up or forgot to convert one of the calls.

Comment (cameel) [reply test/externalTests/common.sh]: Renamed to `FORCE-ABI-V2`

Comment (bshastry) [reply test/externalTests/common.sh]: Last question before approving: None of the external tests force ABIv2. If so, do we need a separate flag?

Comment (bshastry) [reply test/externalTests/common.sh]: Clarified on gitter. tl;dr this is legacy behavior which has simply been refactored.

**Review by bshastry — APPROVED** [review]
LGTM if tests pass. Squash perhaps unless keeping the commits is important?

**Review by leonardoalt — APPROVED** [review]

**Comment by cameel** [comment]
Merged.

I'd rather keep the commits unless there's a good reason not to (e.g. they are messy or just random PR fixes that should logically be part of the original commits). I specifically keep such refactoring changes separate because then it's easier to see the reason behind each one. I really wish it was done that way in `solidity-external-tests`. Rebasing them (and understanding conflicts) would be much easier.
